### PR TITLE
Fix missing ipAddress for ports when LB update & release port resource when remove

### PR DESCRIPTION
**Pull request by carolynhu**
https://github.com/rancher/rancher/issues/9752
There are three issues being fixed in this PR:
When we first create LB service, scheduler will allocate an ip address for the ports. 
(1).However, when we update LB, the original ip address assigned for the old ports is missing (2).also, there is no ip allocated for new ports.

https://github.com/rancher/rancher/issues/9879
(3).port resource fail to release when remove

### CODE REVIEW

File: tests/integration-v1/cattletest/core/test_port.py [hunk 1]
```diff
@@ -33,19 +33,19 @@ def test_container_port_create_start(super_client, new_context):
             assert port.publicPort is None
             assert port.protocol == 'tcp'
             assert port.instanceId == c.id
-            assert port.state == 'inactive'
+            assert port.state == 'requested'
```
Comment (alena1108) [inline tests/integration-v1/cattletest/core/test_port.py]: We should figure out why the test is failing; modifying the test so it just passes, is not quite correct.

Comment (carolynhu) [reply tests/integration-v1/cattletest/core/test_port.py]: @alena1108 just a test to see whether there is only one error with my pr ☺

File: code/iaas/allocator/src/main/java/io/cattle/platform/allocator/dao/impl/AllocatorDaoImpl.java [hunk 2]
```diff
@@ -591,6 +596,35 @@ private void updateInstancePorts(List<Map<String, Object>> dataList) {
         return;
     }
     
+    @SuppressWarnings("unchecked")
+	private void updateInstancePortsForLoadBalancer(List<Map<String, Object>> dataList, List<Port> newPorts) {
```
Comment (alena1108) [inline code/iaas/allocator/src/main/java/io/cattle/platform/allocator/dao/impl/AllocatorDaoImpl.java]: I'd change the name for this method. It should not be load balancer specific. Why don't we just put all the logic to updateInstanceForNewPorts

File: code/iaas/allocator/src/main/java/io/cattle/platform/allocator/dao/AllocatorDao.java [hunk 3]
```diff
@@ -62,4 +62,6 @@
     String getAllocatedHostUuid(Volume volume);
 
     Iterator<AllocationCandidate> iteratorHosts(List<String> orderedHostUuids, List<Long> volumes, QueryOptions options);
+
+    void updateInstanceForNewPorts(List<Map<String, Object>> dataList, List<Port> newPorts);
```
Comment (alena1108) [inline code/iaas/allocator/src/main/java/io/cattle/platform/allocator/dao/AllocatorDao.java]: Usually it is not a very good idea to pass a generic field to the method call. Can you see if we can make the "dataList" type more concrete + the name should reflect the meaning, dataList as a name is too generic as well

Comment (alena1108) [reply code/iaas/allocator/src/main/java/io/cattle/platform/allocator/dao/AllocatorDao.java]: Also are we just updating ports in this method, or instance as well? If just ports, the function name should reflect that

File: code/iaas/allocator/src/main/java/io/cattle/platform/allocator/service/impl/AllocatorServiceImpl.java [hunk 4]
```diff
@@ -205,6 +206,16 @@ public void volumeDeallocate(Volume volume) {
     }
 
     @Override
+    public void newPortAllocateForLoadBalancerUpdate(Instance instance, List<Port> newPorts) {
+        callExternalSchedulerForNewPorts(instance, newPorts);
```
Comment (alena1108) [inline code/iaas/allocator/src/main/java/io/cattle/platform/allocator/service/impl/AllocatorServiceImpl.java]: Lets name the methods as:

callExternalScheduleToAllocatePorts
callExternalSchedulerToRemovePorts

File: code/iaas/allocator/src/main/java/io/cattle/platform/allocator/service/AllocatorService.java [hunk 5]
```diff
@@ -19,5 +20,9 @@
 
     void ensureResourcesReservedForStart(Instance instance);
 
+    void newPortAllocateForLoadBalancerUpdate(Instance instance, List<Port> newPorts);
```
Comment (alena1108) [inline code/iaas/allocator/src/main/java/io/cattle/platform/allocator/service/AllocatorService.java]: Lets change the methods names to:

allocatePortsForInstanceUpdate
releasePortsForInstanceUpdate

File: code/iaas/model/src/main/java/io/cattle/platform/core/dao/impl/NetworkDaoImpl.java [hunk 6]
```diff
@@ -165,40 +165,51 @@ public Network getDefaultNetwork(Long accountId) {
...
             if (port.getRemoved() != null || port.getState().equalsIgnoreCase(CommonStatesConstants.REMOVING)) {
                 continue;
             }
-            existingPorts.put(toKey(port), port);
+            existingInstancePorts.put(toKey(port), port);
         }
 
         for (String port : newPortDefs) {
-            PortSpec spec = new PortSpec(port);
+            PortSpec servicePortSpec = new PortSpec(port);
```
Comment (alena1108) [inline code/iaas/model/src/main/java/io/cattle/platform/core/dao/impl/NetworkDaoImpl.java]: Lets not call it "servicePortSpec"; it should be "newPortSpec"

File: code/iaas/model/src/main/java/io/cattle/platform/core/util/PortSpec.java [hunk 7]
```diff
@@ -93,7 +93,7 @@ public Integer getPublicPort() {
         return publicPort;
     }
 
-    public int getPrivatePort() {
+    public Integer getPrivatePort() {
```
Comment (alena1108) [inline code/iaas/model/src/main/java/io/cattle/platform/core/util/PortSpec.java]: why this change? private port can never be null. Public port can though

File: code/iaas/model/src/test/java/io/cattle/platform/core/util/PortSpecTest.java [hunk 8]
```diff
@@ -12,31 +12,31 @@ public void testValidFormats() {
         // format: ip:hostPort:containerPort | ip::containerPort | hostPort:containerPort | containerPort
         PortSpec spec = new PortSpec("80");
 
-        assertEquals(80, spec.getPrivatePort());
+        assertEquals(new Integer(80), spec.getPrivatePort());
```
Comment (alena1108) [inline code/iaas/model/src/test/java/io/cattle/platform/core/util/PortSpecTest.java]: we should revert this change. Private port can never be null, and it should be primitive.

File: code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/process/LoadBalancerServiceUpdatePostListener.java [hunk 9]
```diff
@@ -103,8 +109,18 @@ private void updateServicePorts(Service service, ProcessState state) {
...
-            instance = objectManager.setFields(instance, InstanceConstants.FIELD_PORTS, newPortDefs);
+            List<Port> toUpdate = objectManager.find(Port.class, PORT.INSTANCE_ID, instance.getId(), PORT.REMOVED, null);
+            List<String> toUpdatePortDefs = new ArrayList<>();
+
+            for (Port port : toUpdate) {
+                PortSpec portSpec = new PortSpec(port);
+                toUpdatePortDefs.add(portSpec.toSpec());
+            }
+
+            instance = objectManager.setFields(instance, InstanceConstants.FIELD_PORTS, toUpdatePortDefs);
```
Comment (alena1108) [inline code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/process/LoadBalancerServiceUpdatePostListener.java]: I'd put this call (objectManager.setFields(instance, InstanceConstants.FIELD_PORTS, toUpdatePortDefs) to corresponding allocatorDao method. So the listener calls allocator service, allocator service calls external scheduler to allocate the ip address, then Dao is called to persist the ports + update the instance.

File: code/iaas/model/src/main/java/io/cattle/platform/core/util/PortSpec.java (hunk 7 above)
Comment (carolynhu) [reply code/iaas/model/src/main/java/io/cattle/platform/core/util/PortSpec.java]: then, here https://github.com/carolynhu/cattle/blob/d688ecb308f04ac730cbdd456d751bb1f2663d6f/code/iaas/model/src/main/java/io/cattle/platform/core/dao/impl/NetworkDaoImpl.java#L182
should be check "instancePortSpec.getPrivatePort() == servicePortSpec.getPrivatePort()" not equals

File: code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/process/LoadBalancerServiceUpdatePostListener.java [hunk 10]
```diff
@@ -117,6 +133,7 @@ private void updateServicePorts(Service service, ProcessState state) {
             for (Port port : toRemove) {
                 deactivateThenRemove(port, new HashMap<String, Object>());
             }
+            allocatorService.portResourceReleasedForLoadBalancerUpdate(instance, toRemove);
```
Comment (alena1108) [inline code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/process/LoadBalancerServiceUpdatePostListener.java]: Comment above applies for release as well.

**Comment by alena1108** [comment]
@carolynhu could you also squash the commits into 1

File: code/iaas/allocator/src/main/java/io/cattle/platform/allocator/dao/AllocatorDao.java (hunk 3 above)
Comment (carolynhu) [reply code/iaas/allocator/src/main/java/io/cattle/platform/allocator/dao/AllocatorDao.java]: The name can be changed such as "allocatedDataList", since this dataList is like: [{allocatedIPs=[{allocatedIP=192.168.60.60, privatePort=11, protocol=tcp, publicPort=11}, {allocatedIP=192.168.60.60, privatePort=22, protocol=tcp, publicPort=22}], instanceID=494}] which is hard to make it to a specific data type

**Comment by alena1108** [comment]
LGTM. Drone TF is unrelated.
